aten/src/ATen/Quantizer.h

```diff
@@ -0,0 +1,199 @@
+  UniformQuantizer(QScheme qscheme): Quantizer(qscheme) {}
+  virtual ~UniformQuantizer();
+};
+
+/**
+ * NonUniformQuantizer is the parent class for all non-uniform quantizers.
+ * These quantization scheme may map float value non-uniformly to the quantized
+ * value. K-means quantization is a representative example in this category.
+ */
+struct CAFFE2_API NonUniformQuantizer : public Quantizer {
+  NonUniformQuantizer() {}
+  NonUniformQuantizer(QScheme qscheme): Quantizer(qscheme) {}
+  virtual ~NonUniformQuantizer();
+
+};
+
+// There is also StochasticQuantizer which is uniform but not affine
+
+/**
+ * AffineQuantizer uses affine transformation to do quantization.
+ *
+ * For quantize:
+ * Y = clamp((X * scale + zero_point, min, max)
+ * For dequantize:
+ * X = (Y - zero_point) / scale
+ */
+struct CAFFE2_API AffineQuantizer : public UniformQuantizer {
+  AffineQuantizer() {}
+  AffineQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~AffineQuantizer();
+};
+
+/**
+ * SymmetricQuantizer is similar to AffineQuantizer except that it
+ * does not have zero_point
+ *
+ * For quantize:
+ * Y = clamp(X * scale, min, max)
+ * For dequantize:
+ * X = Y / scale
+ */
+struct CAFFE2_API SymmetricQuantizer : public UniformQuantizer {
+  SymmetricQuantizer() {}
+  SymmetricQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~SymmetricQuantizer();
+
+};
+
+/**
+ * PerLayerSymmetricQuantizer stores a single scale number which is
+ * used for quantizing all the values in the given Tensor
+ */
+struct CAFFE2_API PerLayerSymmetricQuantizer: public SymmetricQuantizer {
+  PerLayerSymmetricQuantizer() {}
+  PerLayerSymmetricQuantizer(float scale) : SymmetricQuantizer(kPerLayerSymmetric), scale_(scale) {}
+  virtual ~PerLayerSymmetricQuantizer();
+  float scale_{1.0};
+};
+
+/**
+ * PerChannelSymmetricQuantizer stores a vector of scale number and
+ * applys symmetric quantization using different scales.
+ */
+struct CAFFE2_API PerChannelSymmetricQuantizer: public SymmetricQuantizer {
+  PerChannelSymmetricQuantizer() {}
+  PerChannelSymmetricQuantizer(std::vector<float> scales, std::vector<int64_t> channel_axis): SymmetricQuantizer(kPerChannelSymmetric), scales_(scales), channel_axis_(channel_axis) {
+    AT_ASSERT(channel_axis_.size() == 1);
+  }
+  virtual ~PerChannelSymmetricQuantizer();
+
+  std::vector<float> scales_;
+  std::vector<int64_t> channel_axis_;
+};
+
+/**
+ * PerLayerAffineQuantizer stores a scale and a zero_point, which is used for
+ * all the values in the Tensor.
+ */
+struct CAFFE2_API PerLayerAffineQuantizer: public AffineQuantizer{
+  PerLayerAffineQuantizer(float scale, uint8_t zero_point): AffineQuantizer(kPerLayerAffine), scale_(scale), zero_point_(zero_point) {}
+  ~PerLayerAffineQuantizer();
+
+  virtual QTensor quantize(RealTensor tensor);
+  virtual RealTensor dequantize(QTensor tensor);
+
+  float scale() {
+    return scale_;
+  }
+
+  int32_t zero_point() {
+    return zero_point_;
```

Comment:
Correct me if I am wrong: zero_point_ is uint8_t, while the return type for this function is int32_t. There is a static type casting here?

Reply:
right, I changed data type of member but forgot to change this one.